### PR TITLE
[WIP]Add EfficientFormer backbone

### PR DESCRIPTION
**Pull request by zhangyubo0722**
Add UniFormer backbone

### CODE REVIEW

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+import paddle.nn as nn
+from paddle.nn.initializer import TruncatedNormal
+from typing import Dict
+import itertools
+from .vision_transformer import trunc_normal_, to_2tuple, DropPath, zeros_
+
+from ....utils.save_load import load_dygraph_pretrain, load_dygraph_pretrain_from_url
+
+MODEL_URLS = {
+    "EfficientFormer_L1_300epoch":
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:30]: 这个去掉300epoch，链接同，下边的1000epoch的key和value都去掉

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+        self.norm = norm_layer(embed_dim) if norm_layer else nn.Identity()
+
+    def forward(self, x):
+        x = self.proj(x)
+        x = self.norm(x)
+        return x
+
+
+class Flat(nn.Layer):
+    def __init__(self, ):
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:163]: 逗号去掉

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+        self.apply(self._init_weights)
+
+        self.norm1 = nn.BatchNorm2D(hidden_features)
+        self.norm2 = nn.BatchNorm2D(out_features)
+
+    def _init_weights(self, m):
+        if isinstance(m, nn.Conv2D):
+            trunc_normal_(m.weight)
+            if m.bias is not None:
+                nn.initializer.Constant(value=0)(m.bias)
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:242]: 这个直接用zeros_

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+                    value=layer_scale_init_value))
+            self.layer_scale_2 = self.create_parameter(
+                [dim],
+                dtype='float32',
+                default_initializer=nn.initializer.Constant(
+                    value=layer_scale_init_value))
+
+    def forward(self, x):
+        if self.use_layer_scale:
+
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:344]: 去掉空行

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+            blocks.append(
+                Meta3D(
+                    dim,
+                    mlp_ratio=mlp_ratio,
+                    act_layer=act_layer,
+                    norm_layer=norm_layer,
+                    drop=drop_rate,
+                    drop_path=block_dpr,
+                    use_layer_scale=use_layer_scale,
+                    layer_scale_init_value=layer_scale_init_value, ))
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:384]: 去掉逗号

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 6]
```diff
@@ -0,0 +1,563 @@
...
+            blocks.append(
+                Meta4D(
+                    dim,
+                    pool_size=pool_size,
+                    mlp_ratio=mlp_ratio,
+                    act_layer=act_layer,
+                    drop=drop_rate,
+                    drop_path=block_dpr,
+                    use_layer_scale=use_layer_scale,
+                    layer_scale_init_value=layer_scale_init_value, ))
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:395]: 去掉逗号

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 7]
```diff
@@ -0,0 +1,563 @@
...
+                x_out = norm_layer(x)
+                outs.append(x_out)
+        if self.fork_feat:
+            return outs
+        return x
+
+    def forward(self, x):
+        x = self.patch_embed(x)
+        x = self.forward_tokens(x)
+        if self.fork_feat:
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:509]: 这个是做什么的呢

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 8]
```diff
@@ -0,0 +1,563 @@
...
+    "https://paddle-imagenet-models-name.bj.bcebos.com/dygraph/EfficientFormer_L1_1000epoch_pretrained.pdparams",
+    "EfficientFormer_L3":
+    "https://paddle-imagenet-models-name.bj.bcebos.com/dygraph/EfficientFormer_L3_pretrained.pdparams",
+    "EfficientFormer_L7":
+    "https://paddle-imagenet-models-name.bj.bcebos.com/dygraph/EfficientFormer_L7_pretrained.pdparams",
+}
+
+__all__ = list(MODEL_URLS.keys())
+
+EfficientFormer_width = {
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:42]: 这个删除掉，在下边函数定义单独写

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 9]
```diff
@@ -0,0 +1,563 @@
...
+    'l1': [48, 96, 224, 448],
+    'l3': [64, 128, 320, 512],
+    'l7': [96, 192, 384, 768],
+}
+
+EfficientFormer_depth = {
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:48]: 这个删除掉，在下边函数定义单独写

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 10]
```diff
@@ -0,0 +1,563 @@
...
+        for p1 in points:
+            for p2 in points:
+                offset = (abs(p1[0] - p2[0]), abs(p1[1] - p2[1]))
+                if offset not in attention_offsets:
+                    attention_offsets[offset] = len(attention_offsets)
+                idxs.append(attention_offsets[offset])
+        self.attention_biases = self.create_parameter(
+            shape=[num_heads, len(attention_offsets)],
+            dtype='float32',
+            default_initializer=nn.initializer.Constant(value=0.))
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:88]: 这个使用zeros_

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 11]
```diff
@@ -0,0 +1,563 @@
...
+        k = k.transpose(perm=[0, 2, 1, 3])
+        v = v.transpose(perm=[0, 2, 1, 3])
+
+        self.attention_bias_idxs = paddle.flatten(
+            self.attention_bias_idxs, start_axis=0, stop_axis=-1)
+        attn = ((q @k.transpose(
+            (0, 1, 3, 2))) * self.scale + paddle.index_select(
+                self.attention_biases, self.attention_bias_idxs,
+                axis=1).reshape((self.num_heads, self.N, self.N)))
+        attn = nn.Softmax(axis=-1)(attn)
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:111]: 这里用F.softmax

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 12]
```diff
@@ -0,0 +1,563 @@
...
+        x = (attn @v).transpose(perm=[0, 2, 1, 3]).reshape(
+            shape=[B, N, self.dh])
+        x = self.proj(x)
+        return x
+
+
+def stem(in_chs, out_chs):
+    return nn.Sequential(
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:119]: 这里写成两行，如sequential = nn.Sequential(), return sequential

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 13]
```diff
@@ -0,0 +1,563 @@
...
+
+    def __init__(self,
+                 patch_size=16,
+                 stride=16,
+                 padding=0,
+                 in_chans=3,
+                 embed_dim=768,
+                 norm_layer=nn.BatchNorm2D):
+        super().__init__()
+        patch_size = to_2tuple(patch_size)
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:145]: 这里使用map，三个to2tuple吧

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 14]
```diff
@@ -0,0 +1,563 @@
...
+                self.add_module(layer_name, layer)
+        else:
+            # Classifier head
+            self.norm = norm_layer(embed_dims[-1])
+            self.head = nn.Linear(
+                embed_dims[-1], class_num) if class_num > 0 \
+                else nn.Identity()
+        self.apply(self.cls_init_weights)
+
+        self.init_cfg = copy.deepcopy(init_cfg)
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:481]: 这个的必要性是？

File: ppcls/arch/backbone/model_zoo/efficientformer.py [hunk 15]
```diff
@@ -0,0 +1,563 @@
...
+        load_dygraph_pretrain(model, pretrained)
+    else:
+        raise RuntimeError(
+            "pretrained type is not available. Please use `string` or `boolean` type."
+        )
+
+
+def EfficientFormer_L1(pretrained=False, use_ssld=False, **kwargs):
+    model = EfficientFormer(
+        layers=EfficientFormer_depth['l1'],
```
Comment (cuicheng01) [inline ppcls/arch/backbone/model_zoo/efficientformer.py:532]: 这里直接写死就行

**Review by cuicheng01 — COMMENTED** [review]
1.补充文档
2.补充配置
3.补充TIPC
